### ISSUE TITLE
Whonix Time Synchronization Monitor issue when starting new Whonix Disposable while sys-whonix not running

### ISSUE THREAD

**Issue by UndeadDevel** [QubesOS/qubes-issues#8672]
### Qubes OS release
Qubes 4.1.2
Whonix 16

### Brief summary
I've noticed that since I disabled sys-whonix autostart at boot of QubesOS there are issues with the Time Synchronization Monitor (TSM) in the Notification Area of the panel (top panel right side by default). The issues seem to be:
1. it either takes longer for the full initialization to complete (lock with full circle icon) or the notification icon doesn't show the correct status
2. interacting with the notification icon is not possible, because the little popup that appears when clicking on it, immediately disappears again.

This only occurs when starting, e.g. a new Whonix WS DVM while sys-whonix is not running, such that it will automatically start sys-whonix as well. If sys-whonix was already running and the full lock icon displayed then starting new DVMs does not cause this issue; starting sys-whonix alone (e.g. autostart on boot or manually later) does not cause this issue if one waits for "full lock" of the icon before starting sys-whonix-connected VMs.

This can be fixed by running the "Reload Firewall" or "Reload Tor" apps of sys-whonix, though the former seems to be more efficient in terms of time until full lock is displayed.

I've also noticed that when disconnecting Whonix WS DVMs from sys-whonix and reconnecting them to it later, those don't always show up in the TSM again, even though they do regain network access through sys-whonix (as per Qube Manager), though this may be a different problem.

### Steps to reproduce
Launch Whonix WS DVM without sys-whonix running (should autostart together).
Observe and try to interact with the TSM icon in the Notification Area of the panel.

### Expected behavior
Normal TSM functionality

### Actual behavior
Interaction not possible, seems to take much longer for "full lock" to appear, needs fix via e.g. "Reload Firewall" in sys-whonix to work normally again.

**Comment by DemiMarie** [comment]
I am pretty sure this is a Whonix problem, not a Qubes problem.

**Comment by UndeadDevel** [comment]
> I am pretty sure this is a Whonix problem, not a Qubes problem.

Could be, though I don't have a non-Qubes machine right now to verify this on.

Another weird thing about TSM: when turning off compositing in the QubesOS "Window Manager Tweaks" tool the TSM disappears, even if I re-enable compositing shortly thereafter. I have to run the sdwdate-gui app in sys-whonix to make it appear in the panel's notification area again. This does not happen to the Network Manager icon in the panel (which is from sys-net).

**Comment by adrelanos** [comment]
There was a similar recent sdwdate-gui issue which was fixed in all repositories:
https://forums.whonix.org/t/sdwdate-gui-for-qubes-testers-wanted-developers-welcome/5469/21

> it either takes longer for the full initialization to complete (lock with full circle icon)

Taking long time is only an issue if it doesn't correspond with that it's being said in the logs. sdwdate-gui is only the messenger.

> the notification icon doesn't show the correct status

How you know that?

> Another weird thing about TSM: when turning off compositing in the QubesOS "Window Manager Tweaks" tool the TSM disappears, even if I re-enable compositing shortly thereafter.

No idea about that one. Needs a separate bug report.

**Comment by UndeadDevel** [comment]
>No idea about that one. Needs a separate bug report.

Now opened at #8797; I've also opened a new issue regarding another problem mentioned in the OP under #8798.

>There was a similar recent sdwdate-gui issue which was fixed in all repositories:

My Whonix templates are fully up to date, but this/these issue(s) still occur(s).

>How you know that?

I don't, it was just speculation, but in any case it seems this part of the problem (longer sdwdate synchronization) is fixed now in 4.2 with Whonix 17; the other part (lacking usability of the tray widget) has gotten worse, though. I've updated the OP now.

**Comment by adrelanos** [comment]
> interacting with the notification icon is not possible, because the little popup that appears when clicking on it, immediately disappears again.

To attempt to debug this...

Check version.

    dpkg -l | grep sdwdate-gui

Expected output at time of writing:

> ii  sdwdate-gui                                1:9.6-1                         all          Sdwdate Monitor

If not that version then that's the issue. In that case no need to proceed further. If you already have the expected version:

Please close sdwdate-gui. Then start it manually from sys-whonix.

    sdwdate-gui-qubes

(Not sudo.)

The only expected messages are:

> tor_status was called.
> tor_status status: tor_enabled

If there are any errors please post them here.

**Comment by UndeadDevel** [comment]
>dpkg -l | grep sdwdate-gui

```
ii  sdwdate-gui                                1:9.8-1                         all          Sdwdate Monitor
```

So it looks like I'm ahead...I have security-testing updates enabled in the Global Config app, though I'm not sure that this setting is for normal templates and not just for dom0.

>sdwdate-gui-qubes

```
tor_status was called.
tor_status status: tor_enabled
```
Curiously, `sys-whonix` showed up in the widget after that, but not a connected workstation (#8798); when I start the widget from the application menu, `sys-whonix` does not show up there.
Regarding this here issue, it did not return (widget can immediately be interacted with normally), but as it turned out this was likely because I had already "fixed it" by restarting the `sys-whonix` firewall (via app menu); when rebooting `sys-whonix` without applying "the fix" and doing:
```
killall sdwdate-gui-qubes
sdwdate-gui-qubes
```
then I see it having the issue; and again it can be fixed by launching "Reload Firewall"...so it seems that I need to either execute the "Reload Tor" or "Reload Firewall" apps once per `sys-whonix` boot to make the issue go away.

**Comment by adrelanos** [comment]
Correction:

    dpkg -l | grep sdwdate-gui

> ii  sdwdate-gui                                1:9.8-1                         all          Sdwdate Monitor

----

Quote:

> ii  sdwdate-gui                                1:9.8-1                         all          Sdwdate Monitor

So yeah your version is up-to-date.

----

Already started workstations might not show up in a manually restarted sdwdate-gui. This issue is similar to https://github.com/QubesOS/qubes-issues/issues/8798

----

> Regarding this here issue, it did not return

sdwdate-gui-qubes does not return to shell as expected. It is a long running process similar to a daemon (init / systemd unit file). If it would return to shell (terminate / exit) without closing it with the GUI that would actually be a bug.

----

So if you terminate and manually start sdwdate-gui-qubes you cannot reproduce this issue? If yes, that makes the issue harder to debug. In that case, try disable autostart of sdwdate-gui-qubes.

In template:

    sudo mv /etc/xdg/autostart/sdwdate-gui.desktop ~/

Then shutdown Template, restart sys-whonix and manually start sdwdate-gui-qubes there.

    sdwdate-gui-qubes

See if that shows any errors? Can interact normally from the first start then or issues?

issue: "Right click -> context menu goes away too quickly". (Same with left click.) Please lets focus on that issue only and nothing else. For others, please open other tickets if none open yet.

**Comment by UndeadDevel** [comment]
>issue: "Right click -> context menu goes away too quickly". (Same with left click.)

Yes, that's the issue.

>So if you terminate and manually start sdwdate-gui-qubes you cannot reproduce this issue?

Incorrect. I can reproduce it that way on a freshly started `sys-whonix`, just not if I already applied "the fix", i.e. restarting the firewall or tor, which reliably fixes this issue for the current "`sys-whonix`-session".

>See if that shows any errors? Can interact normally from the first start then or issues?

Removing `sdwdate-gui-qubes` from autostart as you suggest and then starting it manually also reproduces the issue, but curiously not always on the first "click", just on many of the subsequent ones. Actually, on more testing I've found that if I click on it long enough, the issue eventually disappears, including in the case when the widget auto-starts. However, it is faster to do "Restart Firewall" from the app menu rather than repeatedly click on the widget for ~15 seconds...also, when "fixed" by clicking on it for a while, it often returns spontaneously after a few minutes and may also go away again spontaneously, while the "Restart Firewall" fix reliably fixes it always. I see no error messages in the terminal or `sudo dmesg` or anything relevant in `sudo journalctl`, I think, though see for yourself (some lines with URLs removed):
<details>
 <summary>journalctl log</summary>

```
Dec 29 20:04:24 host sdwdate[1148]: 2023-12-29 20:04:24 - sdwdate - INFO - sdwdate (Secure Distributed Web Date) started. PID: 1148
Dec 29 20:04:24 host sdwdate[1148]: 2023-12-29 20:04:24 - sdwdate - INFO - sdwdate (Secure Distributed Web Date) started. PID: 1148
Dec 29 20:04:24 host audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=unconfined msg='unit=sdwdate comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? ter>
Dec 29 20:04:24 host systemd[1]: Started sdwdate.service - Secure Distributed Web Date.
Dec 29 20:04:24 host systemd[1]: Reached target multi-user.target - Multi-User System.
Dec 29 20:04:24 host sdwdate[1148]: 2023-12-29 20:04:24 - sdwdate - INFO - Tor socks host: 127.0.0.1 Tor socks port: 9108
Dec 29 20:04:24 host sdwdate[1148]: 2023-12-29 20:04:24 - sdwdate - INFO - Running sdwdate main loop. iteration: 1
Dec 29 20:04:24 host audit[1257]: USER_ACCT pid=1257 uid=1000 auid=4294967295 ses=4294967295 subj=unconfined msg='op=PAM:accounting grantors=pam_faillock,pam_permit acct="user" exe="/usr/bi>
Dec 29 20:04:24 host audit[1257]: USER_CMD pid=1257 uid=1000 auid=4294967295 ses=4294967295 subj=unconfined msg='cwd="/home/user" cmd="/usr/libexec/security-misc/panic-on-oops" exe="/usr/bi>
Dec 29 20:04:24 host audit[1257]: CRED_REFR pid=1257 uid=1000 auid=4294967295 ses=4294967295 subj=unconfined msg='op=PAM:setcred grantors=pam_faillock,pam_wheel,pam_permit acct="root" exe=">
Dec 29 20:04:24 host sudo[1257]:     user : PWD=/home/user ; USER=root ; COMMAND=/usr/libexec/security-misc/panic-on-oops
Dec 29 20:04:24 host audit[1257]: USER_START pid=1257 uid=1000 auid=4294967295 ses=4294967295 subj=unconfined msg='op=PAM:session_open grantors=pam_limits,pam_permit,pam_unix,pam_tmpdir acc>
Dec 29 20:04:24 host audit[1257]: USER_END pid=1257 uid=1000 auid=4294967295 ses=4294967295 subj=unconfined msg='op=PAM:session_close grantors=pam_limits,pam_permit,pam_unix,pam_tmpdir acct>
Dec 29 20:04:24 host audit[1257]: CRED_DISP pid=1257 uid=1000 auid=4294967295 ses=4294967295 subj=unconfined msg='op=PAM:setcred grantors=pam_faillock,pam_wheel,pam_permit acct="root" exe=">
Dec 29 20:04:24 host systemd[1]: Starting systemd-update-utmp-runlevel.service - Record Runlevel Change in UTMP...
Dec 29 20:04:24 host sudo[1257]: pam_unix(sudo:session): session opened for user root(uid=0) by (uid=1000)
Dec 29 20:04:24 host sudo[1257]: pam_unix(sudo:session): session closed for user root
Dec 29 20:04:24 host audit[1253]: SYSTEM_RUNLEVEL pid=1253 uid=0 auid=4294967295 ses=4294967295 subj=unconfined msg='old-level=N new-level=3 comm="systemd-update-utmp" exe="/usr/lib/systemd>
Dec 29 20:04:24 host systemd[1]: systemd-update-utmp-runlevel.service: Deactivated successfully.
Dec 29 20:04:24 host systemd[1]: Finished systemd-update-utmp-runlevel.service - Record Runlevel Change in UTMP.
Dec 29 20:04:24 host audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=unconfined msg='unit=systemd-update-utmp-runlevel comm="systemd" exe="/usr/lib/systemd/systemd" >
Dec 29 20:04:24 host systemd[1]: Startup finished in 1.013s (kernel) + 2.836s (userspace) = 3.849s.
Dec 29 20:04:24 host audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=unconfined msg='unit=systemd-update-utmp-runlevel comm="systemd" exe="/usr/lib/systemd/systemd" h>
Dec 29 20:04:24 host qubes.VMShell-dom0[1288]: 
Dec 29 20:04:24 host qubes.VMShell-dom0[1288]: (xfce4-terminal:1291): Gdk-CRITICAL **: 20:04:24.497: gdk_atom_intern: assertion 'atom_name != NULL' failed
Dec 29 20:04:24 host qubes.VMShell-dom0[1288]: 
Dec 29 20:04:24 host qubes.VMShell-dom0[1288]: (xfce4-terminal:1291): dbind-WARNING **: 20:04:24.502: AT-SPI: Error retrieving accessibility bus address: org.freedesktop.DBus.Error.ServiceU>
Dec 29 20:04:24 host dbus-daemon[1123]: [session uid=1000 pid=1121] Activating service name='org.gtk.vfs.Daemon' requested by ':1.5' (uid=1000 pid=1291 comm="xfce4-terminal")
Dec 29 20:04:24 host dbus-daemon[1123]: [session uid=1000 pid=1121] Successfully activated service 'org.gtk.vfs.Daemon'
Dec 29 20:04:24 host qubes.VMShell-dom0[1288]: Failed to connect to session manager: Failed to connect to the session manager: SESSION_MANAGER environment variable not defined
Dec 29 20:04:24 host vanguards[946]: NOTICE[Fri Dec 29 20:04:24 2023]: Vanguards 0.3.1 connected to Tor 0.4.7.16 using stem 1.8.1
Dec 29 20:04:24 host sdwdate[1148]: 2023-12-29 20:04:24 - sdwdate - INFO - PREPARATION: running onion-time-pre-script...
Dec 29 20:04:24 host sdwdate[1148]: 2023-12-29 20:04:24 - sdwdate - INFO -
Dec 29 20:04:24 host sdwdate[1148]: __ ### START: ### /usr/libexec/helper-scripts/onion-time-pre-script
Dec 29 20:04:24 host sdwdate[1148]: __ Status: First run after boot. (Creating file '/run/sdwdate/onion-time-script-after-boot'.)
Dec 29 20:04:24 host sdwdate[1148]: __ Static Time Sanity Check: Within minimum time 'Mon Jun 12 00:00:00 UTC 2023' and expiration timestamp 'Tue May 17 10:00:00 UTC 2033', ok.
Dec 29 20:04:24 host sdwdate[1148]: __ Tor reports: NOTICE BOOTSTRAP PROGRESS=14 TAG=handshake SUMMARY="Handshaking with a relay"
Dec 29 20:04:24 host sdwdate[1148]: __ Tor circuit: not established
Dec 29 20:04:24 host sdwdate[1148]: __ Tor Consensus Time Sanity Check: Clock within consensus parameters consensus/valid-after 2023-12-29 19:00:00 and consensus/valid-until 2023-12-29 22:0>
Dec 29 20:04:24 host sdwdate[1148]: __ Conclusion: No Tor circuit established yet.
Dec 29 20:04:24 host sdwdate[1148]: __ ### END: ### Exiting with exit_code '2' indicating 'wait, show busy icon and retry.'.
Dec 29 20:04:24 host sdwdate[1148]: 2023-12-29 20:04:24 - sdwdate - INFO - PREPARATION RESULT: onion-time-pre-script recommended to wait. Consider running systemcheck for more information.
Dec 29 20:04:24 host sdwdate[1148]: 2023-12-29 20:04:24 - sdwdate - INFO -
Dec 29 20:04:26 host sdwdate[1148]: 2023-12-29 20:04:26 - sdwdate - INFO - PREPARATION: running onion-time-pre-script...
Dec 29 20:04:26 host sdwdate[1148]: __ Status: Subsequent run after boot.
Dec 29 20:04:26 host sdwdate[1148]: __ Static Time Sanity Check: Within minimum time 'Mon Jun 12 00:00:00 UTC 2023' and expiration timestamp 'Tue May 17 10:00:00 UTC 2033', ok.
Dec 29 20:04:26 host sdwdate[1148]: __ Tor reports: NOTICE BOOTSTRAP PROGRESS=95 TAG=circuit_create SUMMARY="Establishing a Tor circuit"
Dec 29 20:04:26 host sdwdate[1148]: __ Tor circuit: not established
Dec 29 20:04:26 host sdwdate[1148]: __ Tor Consensus Time Sanity Check: Clock within consensus parameters consensus/valid-after 2023-12-29 19:00:00 and consensus/valid-until 2023-12-29 22:0>
Dec 29 20:04:26 host sdwdate[1148]: __ Conclusion: No Tor circuit established yet.
Dec 29 20:04:26 host sdwdate[1148]: __ ### END: ### Exiting with exit_code '2' indicating 'wait, show busy icon and retry.'.
Dec 29 20:04:26 host sdwdate[1148]: 2023-12-29 20:04:26 - sdwdate - INFO - PREPARATION RESULT: onion-time-pre-script recommended to wait. Consider running systemcheck for more information.
Dec 29 20:04:26 host sdwdate[1148]: 2023-12-29 20:04:26 - sdwdate - INFO -
Dec 29 20:04:27 host sdwdate[1148]: 2023-12-29 20:04:27 - sdwdate - INFO - PREPARATION:
Dec 29 20:04:27 host sdwdate[1148]: 2023-12-29 20:04:27 - sdwdate - INFO - __ ### START: ### /usr/libexec/helper-scripts/onion-time-pre-script
Dec 29 20:04:27 host sdwdate[1148]: __ Status: Subsequent run after boot.
Dec 29 20:04:27 host sdwdate[1148]: __ Static Time Sanity Check: Within minimum time 'Mon Jun 12 00:00:00 UTC 2023' and expiration timestamp 'Tue May 17 10:00:00 UTC 2033', ok.
Dec 29 20:04:27 host sdwdate[1148]: __ Tor reports: NOTICE BOOTSTRAP PROGRESS=100 TAG=done SUMMARY="Done"
Dec 29 20:04:27 host sdwdate[1148]: __ Tor circuit: established
Dec 29 20:04:27 host sdwdate[1148]: __ Tor Consensus Time Sanity Check: Clock within consensus parameters consensus/valid-after 2023-12-29 19:00:00 and consensus/valid-until 2023-12-29 22:0>
Dec 29 20:04:27 host sdwdate[1148]: __ Conclusion: Tor already reports circuit established.
Dec 29 20:04:27 host sdwdate[1148]: __ ### END: ### Exiting with exit_code '0' indicating 'success'.
Dec 29 20:04:27 host sdwdate[1148]: 2023-12-29 20:04:27 - sdwdate - INFO - PREPARATION RESULT: SUCCESS.
Dec 29 20:04:27 host sdwdate[1148]: 2023-12-29 20:04:27 - sdwdate - INFO -
Dec 29 20:04:27 host sdwdate[1148]: 2023-12-29 20:04:27 - sdwdate - INFO - Initial time fetching in progress...
Dec 29 20:04:27 host sdwdate[1148]: 2023-12-29 20:04:27 - sdwdate - INFO - Running sdwdate fetch loop. iteration: 1
Dec 29 20:04:27 host sdwdate[1148]: 2023-12-29 20:04:27 - sdwdate - INFO - pool 0: pool_size: 21 url_index: 15 already_picked_number: 1 already_picked_index: [15]
Dec 29 20:04:27 host sdwdate[1148]: 2023-12-29 20:04:27 - sdwdate - INFO - pool 1: pool_size: 12 url_index: 8 already_picked_number: 1 already_picked_index: [8]
Dec 29 20:04:27 host sdwdate[1148]: 2023-12-29 20:04:27 - sdwdate - INFO - pool 2: pool_size: 22 url_index: 6 already_picked_number: 1 already_picked_index: [6]
Dec 29 20:04:27 host sdwdate[1148]: remote_times.py: url_to_unixtime_command (s):
Dec 29 20:04:28 host qubes.VMShell-dom0[1288]: 
Dec 29 20:04:28 host qubes.VMShell-dom0[1288]: (xfce4-terminal:1291): Gdk-WARNING **: 20:04:28.389: ../../../gdk/x11/gdkproperty-x11.c:224 invalid X atom: 1002405094
```

</details>

**Comment by adrelanos** [comment]
Question for developers:
Thoughts on why https://github.com/Kicksecure/sdwdate-gui/blob/master/usr/lib/python3/dist-packages/sdwdate_gui/sdwdate_gui_qubes.py might look up the GUI (during startup)?

**Comment by adrelanos** [comment]
Difficult one. Since I cannot reproduce this locally, I am only guessing what that could be.

There used to be some cases where calls to `qubesdb-read`, `qrexec-client-vm` were not protected by timeouts. So if these were slow or broken then these might have caused the GUI to lock up. These calls are now protected by a ~ 5 seconds timeout.

So if we're lucky, instead of a non-reactive (frozen) sdwdate-gui systray, we'll now get error messages which if seen could be assigned the component which is the root cause.

The improved sdwdate-gui version has been uploaded to all Whonix 17 repositories just now.

**Comment by UndeadDevel** [comment]
I received the update; the issue persists, but in `sudo journalctl` I see something that looks new:
```
Dec 29 21:10:28 host enable-firewall[1490]: 2023-12-29 21:10:28 - /usr/bin/whonix-gateway-firewall - OK: Whonix firewall loaded.
Dec 29 21:10:28 host systemd[1]: Finished whonix-firewall.service - Whonix firewall loader.
Dec 29 21:10:28 host audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=unconfined msg='unit=whonix-firewall comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? ad>
Dec 29 21:10:28 host systemd[1]: Starting qubes-whonix-network.service - sets up Qubes-Whonix-Gateway external and internal network interface...
Dec 29 21:10:28 host network-proxy-setup[1624]: + source /usr/lib/qubes-whonix/utility_functions.sh
Dec 29 21:10:28 host network-proxy-setup[1624]: ++ systemctl -q is-active qubes-updates-proxy-forwarder.socket
Dec 29 21:10:28 host network-proxy-setup[1624]: ++ PROXY_SERVER=http://-censored-/
Dec 29 21:10:28 host network-proxy-setup[1624]: ++ PROXY_META='<meta name="application-name" content="tor proxy"/>'
Dec 29 21:10:28 host network-proxy-setup[1624]: + test -f /run/qubes/this-is-templatevm
Dec 29 21:10:28 host network-proxy-setup[1624]: + '[' '!' -e /usr/share/anon-gw-base-files/gateway ']'
Dec 29 21:10:28 host network-proxy-setup[1624]: + /sbin/ethtool -K eth0 sg off
Dec 29 21:10:28 host network-proxy-setup[1624]: + INTERFACE=eth1
Dec 29 21:10:28 host network-proxy-setup[1624]: + ip link show eth1
Dec 29 21:10:28 host network-proxy-setup[1627]: 3: eth1: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
Dec 29 21:10:28 host network-proxy-setup[1627]:     link/ether -censored- 
Dec 29 21:10:28 host network-proxy-setup[1624]: + exit 0
Dec 29 21:10:28 host systemd[1]: Finished qubes-whonix-network.service - sets up Qubes-Whonix-Gateway external and internal network interface.
Dec 29 21:10:28 host audit[1]: SERVICE_START pid=1 uid=0 auid=4294967295 ses=4294967295 subj=unconfined msg='unit=qubes-whonix-network comm="systemd" exe="/usr/lib/systemd/systemd" hostname>
Dec 29 21:10:31 host qubes-gui[927]: XGetWindowAttributes for 0xc00019 failed in handle_create, ret=0x0
```

Is it related?

**Comment by adrelanos** [comment]
What's new? XGetWindowAttributes? Very old one.

* https://github.com/QubesOS/qubes-issues/issues/2879

**Comment by UndeadDevel** [comment]
Oh ok; yeah I just didn't notice that entire block before in the logs and I thought maybe that error / warning had something to do with this here issue.

**Comment by adrelanos** [comment]
Do you have (minimal) python skills?

Next step would be editing `/usr/lib/python3/dist-packages/sdwdate_gui/sdwdate_gui_qubes.py`, adding a lot of

    print("1")
    print("2")
    print("3")

below lots of lines to figure out where the code hangs during startup. Every few (logical) lines/blocks. Not hard. Just make sure to use the same indentation style, which means in this context the same number of spaces as the line before insert of the print statement.

----

Alternative instruction:

To debug the `sdwdate_gui_qubes.py` file and identify where it hangs during startup, you can insert print statements at various points in the code. Here's how to do it:

Open the file located at `/usr/lib/python3/dist-packages/sdwdate_gui/sdwdate_gui_qubes.py` using a text editor, preferably one that supports Python syntax highlighting but the latter is optional.

    sudoedit /usr/lib/python3/dist-packages/sdwdate_gui/sdwdate_gui_qubes.py

Throughout the code, add print statements like `print("1")`, `print("2")`, and `print("3")`. It's important to add these statements in logical blocks or after every few lines of code.

Mind indentation style. Python relies on indentation to define code blocks. Always use the same number of spaces for indentation as the preceding line where you're inserting the print statement.

After adding a few print statements, save the file and restart the program to check where it hangs. This will help pinpoint the problematic area.

By methodically adding and testing these print statements, you can effectively debug the code and find the cause of the issue.

**Comment by adrelanos** [comment]
I could have briefly looked at your github profile. My instructions are most likely needlessly verbose. There you did way more complex stuff before. Somehow didn't come to mind. Sorry.

**Comment by UndeadDevel** [comment]
Yeah I'll try that next, thanks; probably will be next week, however.

**Comment by UndeadDevel** [comment]
I don't think anything is wrong with that python file; rather with the way the widget interacts with the dom0 Xfce environment:
If I uncheck "Activate focus stealing prevention" in dom0 "Window Manager Tweaks" ("Focus" tab) then the issue disappears; if I check it again the issue reappears...so it seems that the context menu is treated as a "pop up" window by the xfce window manager and the focus stealing prevention kicks in, giving focus back the previous window.
This does not happen for the Network Manager tray widget, however; so this issue may be connected to #8797 perhaps, as `sdwdate-gui` suffers from it as well, unlike any other tray widget.

What's really mysterious to me is why "Reload Firewall" also fixes the issue (for that session) without me changing the Focus settings.

Anyway, I'm off to celebrate; Happy New Years celebration to all!

**Comment by adrelanos** [comment]
* affects Whonix: yes
* caused by Whonix: no
* diagnosed: Why Whonix is affected, yes but why Qubes focus stealing prevention breaks it or how to fix it, no.
* can this be fixed on the Whonix side: Probably not.
* Qubes bug or limitation: Probably yes.
* Can I help with this any further: Probably not.

suggested new issue title:

> Qubes Focus Stealing Prevention Breaks `sdwdate-gui` systray window updates and popups

**Comment by UndeadDevel** [comment]
I've just discovered that actually opening any `sys-whonix` window will do, including a regular terminal, to fix this issue...the issue remains fixed even after that window is closed again.

**Comment by ArrayBolt3** [comment]
Affects 4.3

**Comment by ArrayBolt3** [comment]
This affects more than just sdwdate-gui; it also affects my ability to toggle Element's window visibility by clicking the Element icon in the system tray. When I click the task bar button while the window is visible, it's supposed to disappear, but it usually won't until I press the button repeatedly. After pressing it enough times, sdwdate-gui's menu also "magically" starts working.
